### PR TITLE
Run IXtextModelListeners on the UI thread

### PR DESCRIPTION
**Pull request by mark-christiaens**
Some of the IXtextModelListeners silently assume that the IDocument that
is associated with the XtextResource is not being changed during the
listener's run.  This patch achieves this by running the
IXtextModelListeners on the UI thread.

eclipse/xtext-eclipse#816

### CODE REVIEW

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/XtextDocument.java [hunk 1]
```diff
@@ -556,6 +556,22 @@ protected XtextResource getState() {
...
+		/**
+		 * Run the model listeners on the UI thread.
+		 * 
+		 * @note Some of the {@link IXtextModelListener}s assume that the {@link XtextDocument} that is associated with the model is not
+		 *       changing while they run. To achieve this, we run the {@link IXtextModelListener}s on the UI thread.
+		 */
+		private void notifyModelListenersOnUiThread() {
+			Display display = PlatformUI.getWorkbench().getDisplay();
+			display.asyncExec(() -> {
+				XtextDocument.this.readOnly(((XtextResource resource) -> {
```
Comment (tivervac) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/model/XtextDocument.java]: Might want to use tryReadOnly here

**Comment by cdietrich** [comment]
please have a look at the test failures

**Comment by mark-christiaens** [comment]
I fixed the test failure of org.eclipse.xtext.ui.tests.editor.model.edit.RealXtextDocumentModifyTest.testModelListenerNotNotified().  
The second failure is still a mystery to me.  It claims to run out of heap.  I ran the test a couple of hundred times on my system but no problems ...  Let's see what the CI tells now.

**Comment by mark-christiaens** [comment]
OK, 1 failure to go.  `org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest.testMarkOccurrencesCrossFile` is still running out of memory.  I have failed to reproduce this on my system.  I'll try running this test through maven instead of inside the Eclipse IDE and see if I learn something.

**Comment by mark-christiaens** [comment]
Strange.  I now ran the failing test on my machine using the Maven flow:
```
JAVA_TOOL_OPTIONS=-Xmx2G mvn --batch-mode -fae -Dmaven.test.failure.ignore=true -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn -Poxygen -Dtest=org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest verify
```
Note that I gave the test 2G to run because I see that this is the setting used by the Jenkins log.

I don't run into problems:
```
-------------------------------------------------------------------------------
Test set: org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest
-------------------------------------------------------------------------------
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5.418 sec - in org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest
testMarkOccurrencesCrossFile(org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest)  Time elapsed: 3.329 sec
testMarkOccurrences(org.eclipse.xtext.ui.tests.editor.occurrences.MarkOccurrencesTest)  Time elapsed: 2.072 sec
```

**Comment by mark-christiaens** [comment]
I modified my past slightly so that it differentiates between two cases:

- When we are already running on the UI thread, continue to do so and immediately run the listeners
- When we are running on a non-UI thread, schedule an asynchronous job to run the listeners on the UI thread.

That seems to have done the trick.

File: org.eclipse.xtext.purexbase.ui/xtend-gen/org/eclipse/xtext/purexbase/ui/PureXbaseUiModule.java [hunk 2]
```diff
@@ -36,7 +36,7 @@
     return PureXbaseUiModule.NoFolding.class;
   }
   
-  public PureXbaseUiModule(final AbstractUIPlugin plugin) {
-    super(plugin);
+  public PureXbaseUiModule(final AbstractUIPlugin arg0) {
```
Comment (szarnekow) [inline org.eclipse.xtext.purexbase.ui/xtend-gen/org/eclipse/xtext/purexbase/ui/PureXbaseUiModule.java]: This should probably be reverted.

File: org.eclipse.xtext.ui/xtend-gen/org/eclipse/xtext/ui/editor/bracketmatching/BracePairMatcher.java [hunk 3]
```diff
@@ -57,19 +57,19 @@ public void setBracePairProvider(final IBracePairProvider bracePairProvider) {
     return _xblockexpression;
   }
   
+  public void dispose() {
```
Comment (szarnekow) [inline org.eclipse.xtext.ui/xtend-gen/org/eclipse/xtext/ui/editor/bracketmatching/BracePairMatcher.java]: This looks like an unnecessary change, too.

File: org.eclipse.xtext.ui/xtend-gen/org/eclipse/xtext/ui/editor/reconciler/CancelIndicatorBasedProgressMonitor.java [hunk 4]
```diff
@@ -36,31 +36,31 @@ public boolean isCanceled() {
     return this.cancelIndicator.isCanceled();
   }
   
-  public void beginTask(final String name, final int totalWork) {
-    this.doDelegate.beginTask(name, totalWork);
+  public void beginTask(final String arg0, final int arg1) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ui/xtend-gen/org/eclipse/xtext/ui/editor/reconciler/CancelIndicatorBasedProgressMonitor.java]: And so does this.

**Comment by mark-christiaens** [comment]
I think that I've done everything required.  Can this be merged?  (I'm unfamiliar with the Eclipse process for this final step)

**Comment by holgerschill** [comment]
LGTM
